File: Runtime/StockFlowTransformer.Normalize.vb
'*********************************************************************************************
' STSimStockFlow: A SyncroSim Module for the ST-Sim Stocks and Flows Add-In.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.STSim
Imports System.Globalization

Partial Class StockFlowTransformer

    ''' <summary>
    ''' Normalizes the output options data feed
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub NormalizeOutputOptions()

        Dim drrc As DataRow = Me.ResultScenario.GetDataSheet("STSim_RunControl").GetDataRow()
        Dim MaxTimestep As Integer = CInt(drrc("MaximumTimestep"))
        Dim dsoo As DataSheet = Me.ResultScenario.GetDataSheet(DATASHEET_OO_NAME)
        Dim droo As DataRow = dsoo.GetDataRow()

        If (droo Is Nothing) Then

            droo = dsoo.GetData().NewRow()
            dsoo.GetData().Rows.Add(droo)

        End If

        If (Not AnyOutputOptionsSelected()) Then

            DataTableUtilities.SetRowValue(droo, DATASHEET_OO_SUMMARY_OUTPUT_ST_COLUMN_NAME, CInt(True))
            DataTableUtilities.SetRowValue(droo, DATASHEET_OO_SUMMARY_OUTPUT_FL_COLUMN_NAME, CInt(True))
            DataTableUtilities.SetRowValue(droo, DATASHEET_OO_SUMMARY_OUTPUT_ST_TIMESTEPS_COLUMN_NAME, 1)
            DataTableUtilities.SetRowValue(droo, DATASHEET_OO_SUMMARY_OUTPUT_FL_TIMESTEPS_COLUMN_NAME, 1)

            Me.RecordStatus(StatusType.Information, NO_SUMMARY_OUTPUT_OPTIONS_INFORMATION)

        End If

        Me.ValidateTimesteps(droo, DATASHEET_OO_SUMMARY_OUTPUT_ST_COLUMN_NAME, DATASHEET_OO_SUMMARY_OUTPUT_ST_TIMESTEPS_COLUMN_NAME, "Summary stocks", MaxTimestep)
        Me.ValidateTimesteps(droo, DATASHEET_OO_SUMMARY_OUTPUT_FL_COLUMN_NAME, DATASHEET_OO_SUMMARY_OUTPUT_FL_TIMESTEPS_COLUMN_NAME, "Summary flows", MaxTimestep)
        Me.ValidateTimesteps(droo, DATASHEET_OO_SPATIAL_OUTPUT_ST_COLUMN_NAME, DATASHEET_OO_SPATIAL_OUTPUT_ST_TIMESTEPS_COLUMN_NAME, "Spatial stocks", MaxTimestep)
        Me.ValidateTimesteps(droo, DATASHEET_OO_SPATIAL_OUTPUT_FL_COLUMN_NAME, DATASHEET_OO_SPATIAL_OUTPUT_FL_TIMESTEPS_COLUMN_NAME, "Spatial flows", MaxTimestep)

    End Sub

    ''' <summary>
    ''' Validates the timesteps for the specified column name and maximum timestep
    ''' </summary>
    ''' <param name="dr"></param>
    ''' <param name="optionColumnName"></param>
    ''' <param name="timestepsColumnName"></param>
    ''' <param name="timestepsColumnHeaderText"></param>
    ''' <param name="maxTimestep"></param>
    ''' <remarks></remarks>
    Private Sub ValidateTimesteps(
        ByVal dr As DataRow,
        ByVal optionColumnName As String,
        ByVal timestepsColumnName As String,
        ByVal timestepsColumnHeaderText As String,
        ByVal maxTimestep As Integer)

        If (dr(optionColumnName) Is DBNull.Value) Then
            Return
        End If

        If (Not CBool(dr(optionColumnName))) Then
            Return
        End If

        If (dr(timestepsColumnName) Is DBNull.Value) Then

            Dim message As String = String.Format(CultureInfo.InvariantCulture,
                    "Stocks and Flows Timestep value for '{0}' is invalid.  Using default.", timestepsColumnHeaderText)

            Me.RecordStatus(StatusType.Warning, message)
            dr(timestepsColumnName) = 1

            Return

        End If

        Dim val As Integer = CInt(dr(timestepsColumnName))

        If (val > maxTimestep) Then

            Dim message As String = String.Format(CultureInfo.InvariantCulture,
                "Stocks and Flows Timestep value for '{0}' out of range.  Using default.", timestepsColumnHeaderText)

            Me.RecordStatus(StatusType.Warning, message)
            dr(timestepsColumnName) = maxTimestep

            Return

        End If

    End Sub

    Private Sub NormalizeForUserDistributions()

        If (Me.m_STSimTransformer.DistributionProvider.Values.Count > 0) Then

            Dim Expander As New STSimDistributionBaseExpander(Me.m_STSimTransformer.DistributionProvider)
            Me.ExpandFlowMultipliers(Expander)

        End If

    End Sub

    Private Sub ExpandFlowMultipliers(ByVal expander As STSimDistributionBaseExpander)

        If (Me.m_FlowMultipliers.Count > 0) Then

            Dim NewItems As IEnumerable(Of STSimDistributionBase) =
                expander.Expand(Me.m_FlowMultipliers)

            Me.m_FlowMultipliers.Clear()

            For Each t As FlowMultiplier In NewItems
                Me.m_FlowMultipliers.Add(t)
            Next

        End If

    End Sub

End Class

File: src/Runtime/StockFlowTransformer.Normalize.cs
// stsim-stockflow: SyncroSim Add-On Package (to stsim) for integrating stocks and flows into state-and-transition simulation models in ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using System.Linq;
using SyncroSim.Core;
using SyncroSim.STSim;
using System.Globalization;
using System.Collections.Generic;

namespace SyncroSim.STSimStockFlow
{
	internal partial class StockFlowTransformer
	{
		/// <summary>
		/// Normalizes the output options data feed
		/// </summary>
		/// <remarks></remarks>
		private void NormalizeOutputOptions()
		{
			DataRow drrc = this.ResultScenario.GetDataSheet("STSim_RunControl").GetDataRow();
			int MaxTimestep = Convert.ToInt32(drrc["MaximumTimestep"], CultureInfo.InvariantCulture);
			DataSheet dsoo = this.ResultScenario.GetDataSheet(Constants.DATASHEET_OO_NAME);
			DataRow droo = dsoo.GetDataRow();

			if (droo == null)
			{
				droo = dsoo.GetData().NewRow();
				dsoo.GetData().Rows.Add(droo);
			}

			if (!AnyOutputOptionsSelected())
			{
				DataTableUtilities.SetRowValue(droo, Constants.DATASHEET_OO_SUMMARY_OUTPUT_ST_COLUMN_NAME, Booleans.BoolToInt(true));
				DataTableUtilities.SetRowValue(droo, Constants.DATASHEET_OO_SUMMARY_OUTPUT_FL_COLUMN_NAME, Booleans.BoolToInt(true));
				DataTableUtilities.SetRowValue(droo, Constants.DATASHEET_OO_SUMMARY_OUTPUT_ST_TIMESTEPS_COLUMN_NAME, 1);
				DataTableUtilities.SetRowValue(droo, Constants.DATASHEET_OO_SUMMARY_OUTPUT_FL_TIMESTEPS_COLUMN_NAME, 1);

				this.RecordStatus(StatusType.Information, Constants.NO_SUMMARY_OUTPUT_OPTIONS_INFORMATION);
			}

			this.ValidateTimesteps(droo, Constants.DATASHEET_OO_SUMMARY_OUTPUT_ST_COLUMN_NAME, Constants.DATASHEET_OO_SUMMARY_OUTPUT_ST_TIMESTEPS_COLUMN_NAME, "Summary stocks", MaxTimestep);
			this.ValidateTimesteps(droo, Constants.DATASHEET_OO_SUMMARY_OUTPUT_FL_COLUMN_NAME, Constants.DATASHEET_OO_SUMMARY_OUTPUT_FL_TIMESTEPS_COLUMN_NAME, "Summary flows", MaxTimestep);
			this.ValidateTimesteps(droo, Constants.DATASHEET_OO_SPATIAL_OUTPUT_ST_COLUMN_NAME, Constants.DATASHEET_OO_SPATIAL_OUTPUT_ST_TIMESTEPS_COLUMN_NAME, "Spatial stocks", MaxTimestep);
			this.ValidateTimesteps(droo, Constants.DATASHEET_OO_SPATIAL_OUTPUT_FL_COLUMN_NAME, Constants.DATASHEET_OO_SPATIAL_OUTPUT_FL_TIMESTEPS_COLUMN_NAME, "Spatial flows", MaxTimestep);
			this.ValidateTimesteps(droo, Constants.DATASHEET_OO_LATERAL_OUTPUT_FL_COLUMN_NAME, Constants.DATASHEET_OO_LATERAL_OUTPUT_FL_TIMESTEPS_COLUMN_NAME, "Lateral flows", MaxTimestep);
        }

		/// <summary>
		/// Validates the timesteps for the specified column name and maximum timestep
		/// </summary>
		/// <param name="dr"></param>
		/// <param name="optionColumnName"></param>
		/// <param name="timestepsColumnName"></param>
		/// <param name="timestepsColumnHeaderText"></param>
		/// <param name="maxTimestep"></param>
		/// <remarks></remarks>
		private void ValidateTimesteps(DataRow dr, string optionColumnName, string timestepsColumnName, string timestepsColumnHeaderText, int maxTimestep)
		{
			if (dr[optionColumnName] == DBNull.Value)
			{
				return;
			}

			if (!Convert.ToBoolean(dr[optionColumnName], CultureInfo.InvariantCulture))
			{
				return;
			}

			if (dr[timestepsColumnName] == DBNull.Value)
			{
				string message = string.Format(CultureInfo.InvariantCulture, 
                    "Stocks and Flows Timestep value for '{0}' is invalid.  Using default.", 
                    timestepsColumnHeaderText);

				this.RecordStatus(StatusType.Warning, message);
				dr[timestepsColumnName] = 1;

				return;
			}

			int val = Convert.ToInt32(dr[timestepsColumnName], CultureInfo.InvariantCulture);

			if (val > maxTimestep)
			{
				string message = string.Format(CultureInfo.InvariantCulture, 
                    "Stocks and Flows Timestep value for '{0}' out of range.  Using default.", 
                    timestepsColumnHeaderText);

				this.RecordStatus(StatusType.Warning, message);
				dr[timestepsColumnName] = maxTimestep;

				return;
			}
		}

		private void NormalizeForUserDistributions()
		{
			if (this.m_STSimTransformer.DistributionProvider.Values.Count > 0)
			{
				STSimDistributionBaseExpander Expander = new STSimDistributionBaseExpander(this.m_STSimTransformer.DistributionProvider);
				this.ExpandFlowMultipliers(Expander);
			}
		}

		private void ExpandFlowMultipliers(STSimDistributionBaseExpander expander)
		{
			if (this.m_FlowMultipliers.Count > 0)
			{
				IEnumerable<STSimDistributionBase> NewItems = expander.Expand(this.m_FlowMultipliers);
				this.m_FlowMultipliers.Clear();

				foreach (FlowMultiplier t in NewItems)
				{
					this.m_FlowMultipliers.Add(t);
				}
			}
		}
	}
}